bestofjs/bestofjs
# Refactor backend tasks iteration logic + admin improvements

## Goal

Refactor the logic to iterate through projects and repos in the backend tasks (`backend` app of the monorepo).
The way it works now : the task runner provides two functions `processProjects` and `processRepos` used to loop through all records of the database, for `projects` and `repos` tables

A lot of the logic is shared between the two helpers but there are differences about how to query data.

Strategy:

- Use an abstract class `ItemProcessor` to share the common feature, basically the loop through all the records, handling the pagination options (used mainly when testing tasks locally from the command line), the concurrency, the throttling interval...
- Extend the abstract class for each types of record, handling how the fetch the IDs of the elements to process
  - `ProjectProcessor` 
  - `RepoProcessor`

This PR also includes also changes related to the tag selection when managing project's tags. 
I wanted the tags to be easy to setup, without having to open the project details page in Edit mode.

At first I introduced [emblor](https://github.com/JaleelB/emblor) package, but I was not satisfied by the integration with shadcn/ui and by the impact on the lock file. Finally, I used a simple drop down menu displayed just after the list of selected tags in the `<ViewTags>` component that is included in both `/projects/[slug]` and `/projects/[slug]/edit` components.

## How to test

Use the dummy tasks `hello-world-*` to check the logic.

For projects:

```sh
bun run apps/backend/src/cli.ts hello-world-projects --logLevel 4 --limit 3 --throttleInterval 1000
```

## Screenshots about tags selection

![image](https://github.com/user-attachments/assets/c60f7afa-6adf-4818-bacd-c340d6e5c885)

![image](https://github.com/user-attachments/assets/7a6d7c99-5b0d-4aea-863e-9e1349001a03)
